### DISCUSSION TITLE
Writing an attributes extension

### DISCUSSION THREAD

**Discussion by gregives** [micromark/micromark#45] [Q&A] [answered]
I've been using [markdown-it](https://www.npmjs.com/package/markdown-it) and [markdown-it-attrs](https://www.npmjs.com/package/markdown-it-attrs) to render Markdown but it has a very large bundle size.

I started writing an extension for micromark to handle attributes similarly but I'm not sure it's possible.

```md
[Link to example website](example.com){data-attribute="something"}
```

Given that a [link's attributes are hard-coded](https://github.com/micromark/micromark/blob/ac44b027357e36694efd2c59babba1b89515e73c/lib/compile/html.mjs#L606), is it impossible to create an extension to support the above syntax?

**Comment by wooorm** [comment, marked as answer]
I generally believe adding extensions is a bad thing.
Directives are somewhat standardized: https://github.com/micromark/micromark-extension-directive. Most parsers support them. Could you use those?

**Reply by gregives** [reply]
I had a play around with directives but I couldn't find a way to add an attribute to a link element directly?

**Reply by wooorm** [reply]
That is correct, but they are more powerful: you can allow any component: `:ComponentName[text]{attributes}`, which could be `:a[text]{href=whatever}`, where you define an `a` component in code to generate an `<a>` element

**Reply by gregives** [reply]
Ah that's a good idea, I didn't think of using a directive to actually render the link! Thank you very much for your help, and thanks for all your work on micromark 😄

**Reply by wooorm** [reply]
They solve *components* in markdown. It’s thinking a bit different, but they are very powerful.
And alternatively: you can also use HTML! `<a href...></a>`. That works in most places too!
And thanks for the kind works!
